File: app/server/backend/routers/pregnancy.py
"""孕程记 - 孕期档案 API。"""

from datetime import date, timedelta
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from backend.database import get_db
from backend.models.pregnancy import Pregnancy
from backend.schemas.common import ApiResponse
from backend.schemas.pregnancy import (
    PregnancyCreate, PregnancyUpdate, PregnancyResponse, GestationalAgeResponse,
)
from backend.services.gestational_calculator import (
    calculate_due_date_from_lmp, calculate_due_date_from_conception,
    calculate_lmp_from_due_date, calculate_gestational_age, get_trimester, days_until_due,
)

router = APIRouter()


@router.post("/pregnancies", response_model=ApiResponse[PregnancyResponse])
async def create_pregnancy(data: PregnancyCreate, db: AsyncSession = Depends(get_db)):
    """创建孕期档案。"""
    # 计算预产期
    lmp_date = None
    due_date_val = None

    if data.last_period_date:
        lmp_date = data.last_period_date
        due_date_val = calculate_due_date_from_lmp(lmp_date)
    elif data.conception_date:
        conception = data.conception_date
        due_date_val = calculate_due_date_from_conception(conception)
        lmp_date = conception - timedelta(days=14)
    elif data.due_date:
        due_date_val = data.due_date
        lmp_date = calculate_lmp_from_due_date(due_date_val)
    else:
        raise HTTPException(status_code=400, detail="必须提供末次月经日期、受精日期或预产期")

    pregnancy = Pregnancy(
        last_period_date=lmp_date,
        conception_date=data.conception_date,
        due_date=due_date_val,
        is_active=1,
        baby_name=data.baby_name,
    )
    db.add(pregnancy)
    await db.flush()
    await db.refresh(pregnancy)

    return ApiResponse(data=PregnancyResponse.model_validate(pregnancy))


@router.get("/pregnancies", response_model=ApiResponse[List[PregnancyResponse]])
async def list_pregnancies(db: AsyncSession = Depends(get_db)):
    """获取孕期列表。"""
    result = await db.execute(select(Pregnancy).order_by(Pregnancy.created_at.desc()))
    pregnancies = result.scalars().all()
    return ApiResponse(data=[PregnancyResponse.model_validate(p) for p in pregnancies])


@router.get("/pregnancies/active", response_model=ApiResponse[PregnancyResponse])
async def get_active_pregnancy(db: AsyncSession = Depends(get_db)):
    """获取当前活跃孕期。"""
    result = await db.execute(select(Pregnancy).where(Pregnancy.is_active == 1))
    pregnancy = result.scalar_one_or_none()
    if not pregnancy:
        return ApiResponse(code=1001, data=None, message="暂无活跃孕期")
    return ApiResponse(data=PregnancyResponse.model_validate(pregnancy))


@router.get("/pregnancies/active/gestational-age", response_model=ApiResponse[GestationalAgeResponse])
async def get_gestational_age(db: AsyncSession = Depends(get_db)):
    """获取当前孕周信息。"""
    result = await db.execute(select(Pregnancy).where(Pregnancy.is_active == 1))
    pregnancy = result.scalar_one_or_none()
    if not pregnancy:
        return ApiResponse(code=1001, data=None, message="暂无活跃孕期")

    lmp = pregnancy.last_period_date
    weeks, days, total = calculate_gestational_age(lmp)
    due = pregnancy.due_date
    days_left = days_until_due(due)

    return ApiResponse(data=GestationalAgeResponse(
        weeks=weeks, days=days, total_days=total,
        due_date=pregnancy.due_date, days_until_due=days_left,
        trimester=get_trimester(weeks),
    ))


@router.get("/pregnancies/{pregnancy_id}", response_model=ApiResponse[PregnancyResponse])
async def get_pregnancy(pregnancy_id: str, db: AsyncSession = Depends(get_db)):
    """获取孕期详情。"""
    pregnancy = await db.get(Pregnancy, pregnancy_id)
    if not pregnancy:
        return ApiResponse(code=1001, data=None, message="孕期档案不存在")
    return ApiResponse(data=PregnancyResponse.model_validate(pregnancy))


@router.put("/pregnancies/{pregnancy_id}", response_model=ApiResponse[PregnancyResponse])
async def update_pregnancy(pregnancy_id: str, data: PregnancyUpdate, db: AsyncSession = Depends(get_db)):
    """更新孕期信息。"""
    pregnancy = await db.get(Pregnancy, pregnancy_id)
    if not pregnancy:
        return ApiResponse(code=1001, data=None, message="孕期档案不存在")

    update_data = data.model_dump(exclude_unset=True)
    for key, value in update_data.items():
        setattr(pregnancy, key, value)

    # 如果修改了末次月经日期，重新计算预产期
    if data.last_period_date is not None:
        lmp = data.last_period_date
        pregnancy.due_date = calculate_due_date_from_lmp(lmp)

    # 如果直接修改了预产期，反推末次月经日期
    if data.due_date is not None and data.last_period_date is None:
        pregnancy.last_period_date = calculate_lmp_from_due_date(data.due_date)

    await db.flush()
    await db.refresh(pregnancy)
    return ApiResponse(data=PregnancyResponse.model_validate(pregnancy))


@router.put("/pregnancies/{pregnancy_id}/activate", response_model=ApiResponse[PregnancyResponse])
async def activate_pregnancy(pregnancy_id: str, db: AsyncSession = Depends(get_db)):
    """设为当前活跃孕期。"""
    from sqlalchemy import update
    # 原子操作：先取消所有活跃
    await db.execute(update(Pregnancy).values(is_active=0))

    pregnancy = await db.get(Pregnancy, pregnancy_id)
    if not pregnancy:
        await db.rollback()
        return ApiResponse(code=1001, data=None, message="孕期档案不存在")

    pregnancy.is_active = 1
    await db.flush()
    await db.refresh(pregnancy)
    return ApiResponse(data=PregnancyResponse.model_validate(pregnancy))

File: server/node/routes/pregnancy.js
const express = require('express');
const router = express.Router();
const db = require('../db');
const gestationalCalc = require('../services/gestational-calculator');

router.post('/pregnancies', (req, res) => {
  try {
    const { last_period_date, conception_date, due_date, baby_name } = req.body;

    if (!last_period_date && !conception_date && !due_date) {
      return res.json({ code: 1001, data: null, message: '必须提供 last_period_date、conception_date 或 due_date 之一' });
    }

    let lmp, calculatedDueDate;

    if (last_period_date) {
      lmp = last_period_date;
      calculatedDueDate = gestationalCalc.calculateDueDateFromLmp(last_period_date);
    } else if (conception_date) {
      lmp = gestationalCalc.calculateLmpFromDueDate(
        gestationalCalc.calculateDueDateFromConception(conception_date)
      );
      calculatedDueDate = gestationalCalc.calculateDueDateFromConception(conception_date);
    } else if (due_date) {
      lmp = gestationalCalc.calculateLmpFromDueDate(due_date);
      calculatedDueDate = due_date;
    }

    const id = db.generateId();
    db.run(
      `INSERT INTO pregnancy (id, last_period_date, conception_date, due_date, baby_name, is_active)
       VALUES (?, ?, ?, ?, ?, 1)`,
      [id, lmp, conception_date || null, calculatedDueDate, baby_name || null]
    );

    const pregnancy = db.queryOne('SELECT * FROM pregnancy WHERE id = ?', [id]);
    res.json({ code: 0, data: pregnancy, message: 'success' });
  } catch (e) {
    res.json({ code: 1001, data: null, message: e.message });
  }
});

router.get('/pregnancies', (req, res) => {
  try {
    const pregnancies = db.queryAll('SELECT * FROM pregnancy ORDER BY created_at DESC');
    res.json({ code: 0, data: pregnancies, message: 'success' });
  } catch (e) {
    res.json({ code: 1001, data: null, message: e.message });
  }
});

router.get('/pregnancies/active', (req, res) => {
  try {
    const active = db.queryOne('SELECT * FROM pregnancy WHERE is_active = 1 LIMIT 1');
    if (!active) {
      return res.json({ code: 1001, data: null, message: '没有活跃的孕期档案' });
    }
    res.json({ code: 0, data: active, message: 'success' });
  } catch (e) {
    res.json({ code: 1001, data: null, message: e.message });
  }
});

router.get('/pregnancies/active/gestational-age', (req, res) => {
  try {
    const active = db.queryOne('SELECT * FROM pregnancy WHERE is_active = 1 LIMIT 1');
    if (!active) {
      return res.json({ code: 1001, data: null, message: '没有活跃的孕期档案' });
    }

    const ga = gestationalCalc.calculateGestationalAge(active.last_period_date);
    const trimester = gestationalCalc.getTrimester(ga.weeks);
    const daysUntil = gestationalCalc.daysUntilDue(active.due_date);

    res.json({
      code: 0,
      data: {
        weeks: ga.weeks,
        days: ga.days,
        total_days: ga.total_days,
        due_date: active.due_date,
        days_until_due: daysUntil,
        trimester
      },
      message: 'success'
    });
  } catch (e) {
    res.json({ code: 1001, data: null, message: e.message });
  }
});

router.get('/pregnancies/:id', (req, res) => {
  try {
    const pregnancy = db.queryOne('SELECT * FROM pregnancy WHERE id = ?', [req.params.id]);
    if (!pregnancy) {
      return res.json({ code: 1001, data: null, message: '孕期档案不存在' });
    }
    res.json({ code: 0, data: pregnancy, message: 'success' });
  } catch (e) {
    res.json({ code: 1001, data: null, message: e.message });
  }
});

router.put('/pregnancies/:id', (req, res) => {
  try {
    const existing = db.queryOne('SELECT * FROM pregnancy WHERE id = ?', [req.params.id]);
    if (!existing) {
      return res.json({ code: 1001, data: null, message: '孕期档案不存在' });
    }

    const { last_period_date, conception_date, due_date, baby_name, is_active } = req.body;
    let updates = [];
    let params = [];

    if (baby_name !== undefined) {
      updates.push('baby_name = ?');
      params.push(baby_name);
    }

    if (is_active !== undefined) {
      updates.push('is_active = ?');
      params.push(is_active ? 1 : 0);
    }

    if (last_period_date !== undefined) {
      updates.push('last_period_date = ?');
      params.push(last_period_date);
      const newDue = gestationalCalc.calculateDueDateFromLmp(last_period_date);
      updates.push('due_date = ?');
      params.push(newDue);
    } else if (conception_date !== undefined) {
      updates.push('conception_date = ?');
      params.push(conception_date);
      const newDue = gestationalCalc.calculateDueDateFromConception(conception_date);
      updates.push('due_date = ?');
      params.push(newDue);
      const newLmp = gestationalCalc.calculateLmpFromDueDate(newDue);
      updates.push('last_period_date = ?');
      params.push(newLmp);
    } else if (due_date !== undefined) {
      updates.push('due_date = ?');
      params.push(due_date);
      const newLmp = gestationalCalc.calculateLmpFromDueDate(due_date);
      updates.push('last_period_date = ?');
      params.push(newLmp);
    }

    if (updates.length === 0) {
      return res.json({ code: 1001, data: null, message: '没有需要更新的字段' });
    }

    updates.push("updated_at = datetime('now')");
    params.push(req.params.id);

    db.run(`UPDATE pregnancy SET ${updates.join(', ')} WHERE id = ?`, params);

    const updated = db.queryOne('SELECT * FROM pregnancy WHERE id = ?', [req.params.id]);
    res.json({ code: 0, data: updated, message: 'success' });
  } catch (e) {
    res.json({ code: 1001, data: null, message: e.message });
  }
});

router.put('/pregnancies/:id/activate', (req, res) => {
  try {
    const existing = db.queryOne('SELECT * FROM pregnancy WHERE id = ?', [req.params.id]);
    if (!existing) {
      return res.json({ code: 1001, data: null, message: '孕期档案不存在' });
    }

    db.run("UPDATE pregnancy SET is_active = 0");
    db.run("UPDATE pregnancy SET is_active = 1, updated_at = datetime('now') WHERE id = ?", [req.params.id]);

    const activated = db.queryOne('SELECT * FROM pregnancy WHERE id = ?', [req.params.id]);
    res.json({ code: 0, data: activated, message: 'success' });
  } catch (e) {
    res.json({ code: 1001, data: null, message: e.message });
  }
});

module.exports = router;
